52/jread/logtimes.py

```diff
@@ -0,0 +1,30 @@
+import os
+import sys
+import re
+
+from datetime import datetime
+
+DATETIME_REGEX = re.compile(r'.*(\d{4})-(\d{2})-(\d{2})T(\d{2}):(\d{2}):(\d{2}).*')
+SHUTDOWN_EVENT = 'Shutdown initiated'
+
+with open(os.path.join(sys.path[0], "messages.log"), "r") as f:
+    loglines = f.readlines()
+
+
+def convert_to_datetime(line):
+    match = re.match(DATETIME_REGEX, line)
+    return datetime(int(match.group(1)), int(match.group(2)), int(match.group(3)), int(match.group(4)),
+                    int(match.group(5)), int(match.group(6))) if match else None
+
+
+def time_between_shutdowns(loglines):
```

Comment:
You can use a list comprehension and built-in functions, similar function from one of our Bites:

```
def time_between_shutdowns(loglines):
    """TODO 2:
       Extract shutdown events ("Shutdown initiated") from loglines and
       calculate the timedelta between the first and last one.
       Return this datetime.timedelta object.
    """
    shutdown_entries = [line for line in loglines if SHUTDOWN_EVENT in line]
    shutdown_times = [convert_to_datetime(event) for event in shutdown_entries]
    return max(shutdown_times) - min(shutdown_times)
```